### PR TITLE
fix(type-utils): treat custom type roots as external

### LINKED ISSUE #6868: Bug: v6 `typeMatchesSpecifier` `PackageSpecifier` should match custom type roots
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### Playground Link

_No response_

### Repro Code

```typescript
typeMatchesSpecifier(
  type,
  {
    from: "package",
    package: "foo"
    name: "Foo",
  },
  parserServices.program)
```

### ESLint Config

_No response_

### tsconfig

```jsonc
{
  "compilerOptions": {
    "typeRoots": ["./typings"],
    "baseUrl": "./",
    "paths": {
      "*": ["typings/*"]
    }
  }
}
```

### Expected Result

Type declarations in `typings` should be treated as external.

### Actual Result

Type declarations in `typings` are treated as local files.

### Additional Info

_No response_

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.0.0-alpha.121` |
| `@typescript-eslint/parser`            | `6.0.0-alpha.121` |
| `@typescript-eslint/scope-manager`     | `6.0.0-alpha.121` |
| `@typescript-eslint/typescript-estree` | `6.0.0-alpha.121` |
| `@typescript-eslint/type-utils`        | `6.0.0-alpha.121` |
| `@typescript-eslint/utils`             | `6.0.0-alpha.121` |
| `TypeScript`                           | `5.0.3` |
| `ESLint`                               | `8.36.0` |
| `node`                                 | `18.15.0` |

### PR DESCRIPTION
**Pull request by RebeccaStevens**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6868
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

### CODE REVIEW

**Comment by RebeccaStevens** [comment]
@marekdedic

**Comment by marekdedic** [comment]
Oh, great! I was going to write in #6866 that I'd like to port that logic to `PackageSpecifier` :D

BTW, there'll be a merge conflict with #6780, but no big deal, they're both tiny changes...

**Comment by marekdedic** [comment]
Adding a test for this would probably be quite convoluted, right?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Yip, similar to the other one, seems reasonable but let's test 👍

**Review by marekdedic — COMMENTED** [review]
I think this is mixing #6861 and #6868 together - I don't know if that's a good thing, a bad thing, or neither :D

File: packages/type-utils/typings/typescript.d.ts [hunk 1]
```diff
@@ -30,4 +30,15 @@ declare module 'typescript' {
      */
     intrinsicName?: string;
   }
+
```
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/type-utils/typings/typescript.d.ts:33]: Yeah for now I think it's fine. If they change that API we should see breakages in the TS version upgrade PR.

File: packages/type-utils/src/TypeOrValueSpecifier.ts [hunk 2]
```diff
@@ -125,7 +125,7 @@ function specifierNameMatches(type: ts.Type, name: string | string[]): boolean {
   if (symbol === undefined) {
     return false;
   }
-  return name.some(item => item === symbol.escapedName);
+  return name.some(item => item === (symbol.escapedName as string));
```
Comment (JoshuaKGoldberg) [inline packages/type-utils/src/TypeOrValueSpecifier.ts:128]:
```diff
<<<<<<< HEAD
  return name.some(item => item === (symbol.escapedName as string));
=======
  return name.some(item => (item as ts.__String) === symbol.escapedName);
>>>>>>> v6
```

lol

**Comment by JoshuaKGoldberg** [comment]
@RebeccaStevens I'm hoping to get through the last bits of v6 work today since we're set to launch on Monday. I think that'll just entail merging `v6` into this PR's branch and adding a few unit tests - which I'm fine to do on my own, and plan on doing in a few hours. Please let me know if you already started and I can hold off 😄

**Comment by JoshuaKGoldberg** [comment]
Hmm, yeah, these unit tests are tough. I ran out of time tonight. Will try again tomorrow.

**Comment by JoshuaKGoldberg** [comment]
😕 I can't figure this out. https://github.com/typescript-eslint/typescript-eslint/commit/43d9d753353fa4342985d459f9a3894c0f153c77 shows a unit test that I would think should populate `program.sourceFileToPackageName`, but it's _`Map(0) {size: 0}`_. Either I've misunderstood how to use the combination of `typeRoots` + `sourceFileToPackageName` _(more likely)_, or that currently-internal API isn't made for this use case _(also plausible)_, or there's a bug with `sourceFileToPackageName` _(less likely)_.

**Comment by JoshuaKGoldberg** [comment]
😕 I can't figure this out. https://github.com/typescript-eslint/typescript-eslint/commit/9b4948056790f764b3864badc57823db82cc528e shows a unit test that I would think should populate `program.sourceFileToPackageName`, but it's _`Map(0) {size: 0}`_. Either I've misunderstood how to use the combination of `typeRoots` + `sourceFileToPackageName` in this unit test _(more likely)_, or that currently-internal API isn't made for this use case _(also plausible)_, or there's a bug with `sourceFileToPackageName` _(less likely)_.

To repro:

1. Put a breakpoint on https://github.com/JoshuaKGoldberg/typescript-eslint/blob/9b4948056790f764b3864badc57823db82cc528e/packages/type-utils/src/TypeOrValueSpecifier.ts#L166
2. Open `TypeOrValueSpecifier.test.ts`
3. Launch the VS Code debugger for _Run currently open type-utils test_ (<kbd>F5</kbd>)
4. Observe that `program.sourceFileToPackageName` is empty, even though earlier in the call stack (https://github.com/JoshuaKGoldberg/typescript-eslint/blob/959819186291a93e17e99e52973a68f3a352c2f5/packages/type-utils/tests/TypeOrValueSpecifier.test.ts#L146), `type` is correctly resolved to `export interface Foo`

**Comment by RebeccaStevens** [comment]
Yeah, I gave up on the test too. I came to the conclusion that internal test API thingy isn't setup to do these sort of tests. I didn't want to open that can of worms of editing that.

If I remember right, local tests I tried using a separate repo passed. I can't really turn those into unit test though.

**Comment by JoshuaKGoldberg** [comment]
This was unintentionally auto-closed when we merged the `v6` branch 🙃 it'll be recreated.

**Review by JoshuaKGoldberg — APPROVED** [review]
Yeah, this is a tricky one. All the existing unit tests still pass, and I couldn't figure out how to add unit tests for the new behavior. It's been sitting for a while... let's ship it. 🤘

Sorry for taking so long!

Cartoon animation of a ship on water

**Pull request merged**